Lib/configparser.py

```diff
@@ -1007,6 +1029,13 @@ def _read(self, fp, fpname):
                     cursect[optname].append(value)
                 # a section header or option header?
                 else:
+                    if self._allow_unnamed_section and cursect is None:
```

Comment:
This function is already unmanageably complex (and complexity checks [disabled in the backport](https://github.com/jaraco/configparser/blob/b244f597d871a83986d5146ab49a4e7c19f8b528/backports/configparser/__init__.py#L1039)). Adding this single boolean parameter has expanded this function from ~120 lines to almost 160 and increases the mccabe cyclometric complexity of this function from 25 to 31 (where a target complexity is <10).

```
 cpython main @ pip-run mccabe -- -m mccabe --min 14 Lib/configparser.py
940:4: 'RawConfigParser._read' 25
 cpython main @ gh pr checkout 117273
Switched to branch 'unnamed-section'
 cpython unnamed-section @ pip-run mccabe -- -m mccabe --min 14 Lib/configparser.py
961:4: 'RawConfigParser._read' 31
```

Moreover, I see lots of logic copied and subtly tweaked.

I'm tempted to say this shouldn't be accepted without first refactoring the implementation to allow for the introduction of this feature to be less invasive.

I can see you've put a lot of work into this, and I don't want to let the perfect be the enemy of the good. Since this change has clean unit tests, I'm okay to say let's accept it but flag an issue to refactor this implementation after.

Reply:
Yes, it has a large complexity to achieve a relatively simple task. A refactor would be wellcome.

Reply:
I've started that work in https://github.com/jaraco/cpython/tree/feature/refactor-configparser-read-complexity, based on this PR, assuming that it's going to go in.